### PR TITLE
TSPS-268 unify PipelineVariableTypesEnum and associated changes

### PR DESCRIPTION
**Pull request by mmorgantaylor**
### Description 

- rename `PipelineInputTypesEnum` to `PipelineVariableTypesEnum`
- use it in base `BasePipelineVariableDefinition` class, so used for both `PipelineInputDefinition` and `PipelineOutputDefinition`
- add a `file_suffix` field to pipeline_input_definitions table in the db, change existing `VCF` types to `FILE`s and add `".vcf.gz"` as the file suffix for those
- refactor enum values `VCF` -> `FILE` and `VCF_ARRAY` -> `FILE_ARRAY`; update the `validate()` method to check the fileSuffix for FILE and FILE_ARRAY types

### Jira Ticket
https://broadworkbench.atlassian.net/browse/TSPS-268

### CODE REVIEW

File: service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java [hunk 1]
```diff
@@ -165,9 +167,10 @@ public String validate(String fieldName, Object value) {
    * the value fails validation, return a descriptive error message in a String.
    *
    * @param fieldName - the name of the field being validated (used to construct error messages)
+   * @paramf fileSuffix - nullable suffix used to validate file types
    * @param value - the value to validate
    */
-  public abstract String validate(String fieldName, Object value);
+  public abstract String validate(String fieldName, String fileSuffix, Object value);
```
Comment (jsotobroad) [inline service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java:173]: hmm i think this signature might be better if we change it to 
```
public abstract String validate(PipelineVariableTypesEnum pipelineVariableTypesEnum, Object value);
```
and then inside the overriding functions we can grab the field name and the suffix if its needed.   What do you think?

File: service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java [hunk 2]
```diff
@@ -43,24 +43,24 @@ public <T> T cast(String fieldName, Object value, TypeReference<T> typeReference
...
-  VCF {
+  FILE {
     @Override
     public <T> T cast(String fieldName, Object value, TypeReference<T> typeReference) {
       return (T) STRING.cast(fieldName, value, new TypeReference<>() {});
     }
 
     @Override
-    public String validate(String fieldName, Object value) {
+    public String validate(String fieldName, String fileSuffix, Object value) {
```
Comment (jsotobroad) [inline service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java:60]: do we think we will have a use case where we allow for multiple suffixes or we dont care about the suffix at all? if so how will this method handle those use cases

File: service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java (hunk 1 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java:173]: hm, the suffix isn't stored in the `PipelineVariableTypesEnum` object, it's in each individual `PipelineInputDefinition` object. if we're confident we're only ever going to call validate on inputs, then we could do 
`public abstract String validate(PipelineInputDefinition pipelineInputDefinition, Object value)`
and the rest like you suggest. what do you think?

File: service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java (hunk 2 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java:60]: great questions. for now I'm ok supporting exactly one suffix (and I should add a test for that in the database test) and adding support for none or multiple if/when we need it. do you have strong feelings otherwise?

File: service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java (hunk 1 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java:173]: if we do pass in the PipelineInputDefinition directly for now, and later decide we want to run validate on pipeline outputs too, then `fileSuffix` will be added to the `BasePipelineVariableDefinition` and we should be able to amend the signature to take an implementation of `BasePipelineVariableDefinition`, right? i.e. doing this can be compatible with validating on outputs if we decide to do that in the future

Comment (jsotobroad) [reply service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java:173]: yup i copy pasted the wrong thign lol.  this is exactly right. PipelineInputDefinition for now and if we want to validate outputs (which we _shouldnt_ have to) then BasePipelineVariableDefinition with a file_suffix being added to the outputs table

File: service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java (hunk 2 above)
Comment (jsotobroad) [reply service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java:60]: hmm  yea i would say for the multiple use case not worth thinking about.  I could go either way on the suffix-less use case.  I dont think any of the possible imputation pipeline inputs we get from users would be a suffix-less use case.  If this is true then yea i would say coding for one speicfic output for every FILE input makes sense

Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java]: cool. technically currently if we have a file input and don't want to check for a suffix, we can have the fileSuffix be an empty string and the validate check will pass. (I just added a test for this)

File: service/src/test/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnumTest.java [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    PipelineInputDefinition stringInputDefinition =
+        new PipelineInputDefinition(
+            1L, commonInputName, "string_input", STRING, null, true, true, null);
+    PipelineInputDefinition fileVcfInputDefinition =
+        new PipelineInputDefinition(
+            1L, commonInputName, "file_vcf_input", FILE, ".vcf.gz", true, true, null);
+    PipelineInputDefinition fileBedInputDefinition =
+        new PipelineInputDefinition(
+            1L, commonInputName, "file_bed_input", FILE, ".bed", true, true, null);
+    PipelineInputDefinition fileNoSuffixInputDefinition =
```
Comment (mmorgantaylor) [inline service/src/test/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnumTest.java:51]: @jsotobroad  here + line 116 is the test I added to show we can define a FILE input without a suffix check by using an empty string as the fileSuffix value.

**Review by jsotobroad — APPROVED** [review]
woohoo awesome

File: service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java (hunk 1 above)
Comment (mmorgantaylor) [reply service/src/main/java/bio/terra/pipelines/common/utils/PipelineVariableTypesEnum.java]: ok perfect. great idea, thanks!
